pkg/common/volume_operations.go

```diff
@@ -426,53 +426,75 @@ func (vo *VolumeOperationsImpl) UpdateCRsAfterVolumeDeletion(ctx context.Context
-	acList := accrd.AvailableCapacityList{}
-	if err = vo.k8sClient.ReadList(ctx, &acList); err != nil {
-		ll.Errorf("Volume was deleted but corresponding AC with SC %s hadn't updated, unable to read list: %v",
-			volumeCR.Spec.StorageClass, err)
-	}
-
-	// search for AC
-	acCR := accrd.AvailableCapacity{}
-	for _, a := range acList.Items {
-		if a.Spec.Location == volumeCR.Spec.Location {
-			acCR = a
-			break
-		}
+
+	// Handle AC, Drive, LVG after volume deletion
+	acCR, err := vo.crHelper.GetACByLocation(volumeCR.Spec.Location)
+	if err != nil {
+		ll.Errorf("AC not found for Volume %s by location %s: %v", volumeCR.Name, volumeCR.Spec.Location, err)
+		return
 	}
-	// AC CR must exist
-	if acCR.Name == "" {
-		ll.Errorf("Unable to find available capacity resource for volume %s", volumeID)
+	driveCR, err := vo.crHelper.GetDriveCRByVolume(&volumeCR)
+	if err != nil {
+		ll.Errorf("Unable to read Drive CR for Volume %s : %v", volumeCR.Name, err)
 		return
 	}
 
-	// for LogicalVolumeGroup SCs we need to delete AC CR when no volumes remain to avoid new allocations since
-	// underlying LogicalVolumeGroup CR is destroying. For other SC just to increase size
-	isDeleted := false
-	lvg := &lvgcrd.LogicalVolumeGroup{}
 	if util.IsStorageClassLVG(volumeCR.Spec.StorageClass) {
-		if err = vo.k8sClient.ReadCR(context.Background(), volumeCR.Spec.Location, "", lvg); err != nil {
-			ll.Errorf("Unable to get LogicalVolumeGroup %s: %v", volumeCR.Spec.Location, err)
+		if err = vo.updateLVGAfterVolumeDeletion(ctx, &volumeCR, driveCR, acCR); err != nil {
 			return
 		}
-
-		if isDeleted, err = vo.deleteLVGIfVolumesNotExistOrUpdate(lvg, volumeCR.Name, &acCR); err != nil {
-			ll.Errorf("Unable to remove volume reference from LogicalVolumeGroup %s: %v", volumeCR.Spec.Location, err)
-		}
 	}
 
-	// if LogicalVolumeGroup wasn't deleted and health of volume is GOOD increase AC size
-	// We don't increase AC size for unhealthy volume to avoid new allocations on top of unhealthy drive/lvg
-	if !isDeleted && volumeCR.Spec.Health == apiV1.HealthGood {
+	if volumeCR.Spec.Health == apiV1.HealthGood {
 		// Increase size of AC using volume size
 		acCR.Spec.Size += volumeCR.Spec.Size
-		if err = vo.k8sClient.UpdateCRWithAttempts(ctx, &acCR, 5); err != nil {
-			ll.Errorf("Unable to update AC %s size: %v", acCR.Name, err)
+		ll.Debugf("Add %d to size of AC %s", volumeCR.Spec.Size, acCR.Name)
+		if err := vo.k8sClient.UpdateCR(ctx, acCR); err != nil {
+			ll.Errorf("Unable to update AC %s CR: %v", acCR.Name, err)
 		}
 	}
 }
 
+func (vo *VolumeOperationsImpl) updateLVGAfterVolumeDeletion(ctx context.Context, volumeCR *volumecrd.Volume,
+	driveCR *drivecrd.Drive, acCR *accrd.AvailableCapacity) error {
+	ll := vo.log.WithFields(logrus.Fields{
+		"method":   "updateLVGAfterVolumeDeletion",
+		"volumeID": volumeCR.Name,
+	})
+
+	lvg := lvgcrd.LogicalVolumeGroup{}
+	if err := vo.k8sClient.ReadCR(ctx, volumeCR.Spec.Location, "", &lvg); err != nil {
```

Comment:
We have already read LVGCR in `vo.crHelper.GetACByLocation`. Could we use that one?

Reply:
sure